lib/IntermediateRepresentation/afforestation.cc

```diff
@@ -0,0 +1,1963 @@
+    const parser::WhereConstruct *, const parser::ForallConstruct *,
+    const parser::CompilerDirective *, const parser::OpenMPConstruct *,
+    const parser::OpenMPEndLoopDirective *>;
+
+struct LinearBeginConstruct : public SumTypeCopyMixin<ConstructVariant> {
+  SUM_TYPE_COPY_MIXIN(LinearBeginConstruct)
+  template<typename T>
+  LinearBeginConstruct(const T &c) : SumTypeCopyMixin{&c} {}
+};
+struct LinearEndConstruct : public SumTypeCopyMixin<ConstructVariant> {
+  SUM_TYPE_COPY_MIXIN(LinearEndConstruct)
+  template<typename T> LinearEndConstruct(const T &c) : SumTypeCopyMixin{&c} {}
+};
+
+template<typename CONSTRUCT>
+const char *GetConstructName(const CONSTRUCT &construct) {
+  return std::visit(
+      common::visitors{
+          [](const parser::AssociateConstruct *) { return "ASSOCIATE"; },
+          [](const parser::BlockConstruct *) { return "BLOCK"; },
+          [](const parser::CaseConstruct *) { return "SELECT CASE"; },
+          [](const parser::ChangeTeamConstruct *) { return "CHANGE TEAM"; },
+          [](const parser::CriticalConstruct *) { return "CRITICAL"; },
+          [](const parser::DoConstruct *) { return "DO"; },
+          [](const parser::IfConstruct *) { return "IF"; },
+          [](const parser::SelectRankConstruct *) { return "SELECT RANK"; },
+          [](const parser::SelectTypeConstruct *) { return "SELECT TYPE"; },
+          [](const parser::WhereConstruct *) { return "WHERE"; },
+          [](const parser::ForallConstruct *) { return "FORALL"; },
+          [](const parser::CompilerDirective *) { return "<directive>"; },
+          [](const parser::OpenMPConstruct *) { return "<open-mp>"; },
+          [](const parser::OpenMPEndLoopDirective *) {
+            return "<open-mp-end-loop>";
+          }},
+      construct.u);
+}
+
+struct AnalysisData {
+  std::map<parser::Label, LinearLabel> labelMap;
+  std::vector<std::tuple<const parser::Name *, LinearLabelRef, LinearLabelRef>>
+      nameStack;
+  LinearLabelBuilder labelBuilder;
+  std::map<const semantics::Symbol *, std::set<parser::Label>> assignMap;
+};
+
+void AddAssign(AnalysisData &ad, const semantics::Symbol *symbol,
+    const parser::Label &label) {
+  ad.assignMap[symbol].insert(label);
+}
+std::vector<LinearLabelRef> GetAssign(
+    AnalysisData &ad, const semantics::Symbol *symbol) {
+  std::vector<LinearLabelRef> result;
+  for (auto lab : ad.assignMap[symbol]) {
+    result.emplace_back(lab);
+  }
+  return result;
+}
+LinearLabel BuildNewLabel(AnalysisData &ad) {
+  return LinearLabel{ad.labelBuilder};
+}
+LinearLabel FetchLabel(AnalysisData &ad, const parser::Label &label) {
+  auto iter{ad.labelMap.find(label)};
+  if (iter == ad.labelMap.end()) {
+    LinearLabel ll{ad.labelBuilder};
+    ll.setReferenced();
+    ad.labelMap.insert({label, ll});
+    return ll;
+  }
+  return iter->second;
+}
+std::tuple<const parser::Name *, LinearLabelRef, LinearLabelRef> FindStack(
```

Comment:
Perhaps a `using` typedef here for the tuple, to make it clear that the result type and stack element types are identical?